### DISCUSSION TITLE
Non uniform coordinate spacing

### DISCUSSION THREAD

**Discussion by kurtsansom** [PrincetonUniversity/athena#399] [Q&A] [answered]
What is the name of  the non-uniform grid spacing being used called? and where does definition come from?

I wrote out the equation defined in `mesh.hpp`  and can kinda see whats going on, but am unsure how it was derived?

I have seen definitions for logarithmic spacing, geometric spacing, and I guess there could be extensions to lagrange polynomials, or any other mapping, but I am not sure how that affects the solution given what is mentioned in the wiki.

the wiki says:
> While Athena++ uses formulae consistent with nonuniform mesh spacing, too large a mesh ratio should not be used; a non-fatal warning is raised if the value is not between 0.9 and 1.1 everywhere. The reconstruction algorithm reduces to first-order if the grid spacing changes too much from cell-to-cell, and moreover, the nonuniform truncation error can produce anomalous reflections of waves.
>
> Note, the geometric spacing is equivalent to logarithmic spacing (uniform dx' in logspace given by the coordinate transformation x' = logb(x1) for some b>0 , b ≠ 1) only for a particular choice of (x1rat, x1min, and x1max).

Examples I would be interested in understanding would be clustering at the edges or clustering at the interior. What complications should I expect when trying to generalize the nonuniform mesh spacing?

**Comment by c-white** [comment, marked as answer]
The grid is always perfectly geometric *in cell widths* (i.e. differences in adjacent face-centered coordinates).

Let's say there are <var>n</var> cells with a ratio of <var>r</var>. Call the faces <var>x<sub>0</sub>, ..., x<sub>n</sub></var> and the cell widths <var>&Delta;<sub>0</sub>, ..., &Delta;<sub>n-1</sub></var>. Then <var>&Delta;<sub>i</sub> = &Delta;<sub>0</sub> r<sup>i</sup></var> and so <var>x<sub>i</sub> = x<sub>0</sub> + &Delta;<sub>0</sub> &sum;<sub>j=0</sub><sup>i-1</sup> r<sup>j</sup></var>. We also have the boundary conditions <var>x<sub>0</sub> = x<sub>min</sub></var> and <var>x<sub>n</sub> = x<sub>max</sub></var>.

To see the connection to logarithmic spacing, suppose the faces were uniformly log-spaced: <var>&ell;(x<sup>i</sup></var>) = &ell;x<sub>min</sub> + i &ell;&Delta;</var> for <var> &ell;&Delta; = (&ell;x<sub>max</sub> - &ell;x<sub>min</sub>) / n</var>. Then <var>x<sup>i</sup> = x<sub>min</sub> &Delta;<sup>i</sup></var>, and we have <var>x<sup>i+1</sup> / x<sup>i</sup></var> is a constant, <var>&Delta;</var>. In this case, though, we also have <var>&Delta;<sub>i</sub> = x<sub>i+1</sub> - x<sub>i</sub> = x<sub>min</sub> &Delta;<sup>i</sup> (&Delta; - 1)</var>, and so <var>&Delta;<sub>i+1</sub> / &Delta;<sub>i</sub></var> is the constant <var>&Delta;</var>. That is, this is also geometric in cell widths with the association <var>&Delta;<sub>0</sub> = x<sub>min</sub> (&Delta; - 1)</var> and <var>r = &Delta;</var>. Thus to get logarithmically spaced faces for given <var>x<sub>min</sub></var>, <var>x<sub>max</sub></var>, and <var>n</var>, choose <var>r = (x<sub>max</sub> / x<sub>min</sub>)<sup>1/n</sup></var>.

If you choose an <var>r</var> less than this value (i.e., closer to <var>1</var>), the grid will be more uniformly spaced (less centrally concentrated). Conversely, larger <var>r</var> results in a more centrally concentrated grid. You could also choose <var>r < 1</var> to make the grid concentrated toward <var>x<sub>max</sub></var>, though I've never seen anyone do this.

All that said, the `x1rat` parameter is there for convenience. Various parts of the code are often optimized for a uniform grid, and reconstruction might be able to do something slightly more accurate in that case, but any grid, even one that is not geometric in cell widths, will work.

**Reply by kurtsansom** [reply]
I found another implementation [gridder ](https://lanl.github.io/gridder/manual.html#) that gives some equations for the geometric spacing. I tested out the function `DefaultMeshGeneratorX1` and the gridder version in python to show they were equivalent to double precision. Then derived an equivalent expression to the one given in `DefaultMeshGeneratorX1` using the gridder definitions and proved to myself that they were the same. I would provide some write up for the equations if anyone is interested. 

@c-white thank you that does provide some insight, so logarithmic spacing is geometrically spaced given the extra definition for `r`

Tangent How do you get equations to show?, quick google search didn't help.

**Reply by c-white** [reply]
I cobbled together the equations in raw html, mostly using `<var>`, `<sup>`, and `<sub>` constructs. It's tedious to say the least, but there aren't many good alternatives. Many years ago github supported mathjax, but abandoned it out of laziness and/or foolishness. Given their responsiveness to users of late, I doubt we'll ever see equation rendering support.

**Reply by felker** [reply]
@kurtsansom would be great if you could write it up and we could add it to the Wiki. I remember debating the geometric vs. logarithmic convention with @dfielding14 in 2019, which led to me adding the paragraph you pasted to the Wiki.

**Reply by kurtsansom** [reply]
I need to go back over the derivation I got. I tried to write it out such that you don't need to store the locations in memory, but haven't been able to get that yet.
